### PR TITLE
qa: test mounting clients when a killed client is not evicted due to laggy OSDs

### PR DESCRIPTION
**Pull request by dparmar18**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: qa/tasks/cephfs/test_client_recovery.py
Comment (joscollin) [inline qa/tasks/cephfs/test_client_recovery.py:904]: Is this always 1 client?

Comment (dparmar18) [reply qa/tasks/cephfs/test_client_recovery.py:904]: no only since we killed one client (you can check we did `self.mount_a.kill()` a few line above), we expect it to be laggy, other client should just be fine

**Comment by dparmar18** [comment]
clients stuck at mounting https://pulpito.ceph.com/dparmar-2024-02-28_07:27:08-fs:functional-main-distro-default-smithi/

**Comment by dparmar18** [comment]
looks like it did mount the client:
```
2024-02-28T08:14:43.639 INFO:tasks.cephfs.fuse_mount:Found client admin socket at /var/run/ceph/ceph-client.0.67436.asok
2024-02-28T08:14:43.639 INFO:teuthology.orchestra.run:Running command with timeout 300
2024-02-28T08:14:43.639 DEBUG:teuthology.orchestra.run.smithi071:> sudo ceph --admin-daemon /var/run/ceph/ceph-client.0.67436.asok status
2024-02-28T08:14:43.775 INFO:teuthology.orchestra.run.smithi071.stdout:{
2024-02-28T08:14:43.775 INFO:teuthology.orchestra.run.smithi071.stdout:    "metadata": {
2024-02-28T08:14:43.775 INFO:teuthology.orchestra.run.smithi071.stdout:        "ceph_sha1": "1fb5d10608dccee73edd6ef8a2c7b804a097c324",
2024-02-28T08:14:43.775 INFO:teuthology.orchestra.run.smithi071.stdout:        "ceph_version": "ceph version 19.0.0-1670-g1fb5d106 (1fb5d10608dccee73edd6ef8a2c7b804a097c324) squid (dev)",
2024-02-28T08:14:43.775 INFO:teuthology.orchestra.run.smithi071.stdout:        "entity_id": "0",
2024-02-28T08:14:43.775 INFO:teuthology.orchestra.run.smithi071.stdout:        "hostname": "smithi071",
2024-02-28T08:14:43.775 INFO:teuthology.orchestra.run.smithi071.stdout:        "mount_point": "/home/ubuntu/cephtest/mnt.0",
2024-02-28T08:14:43.775 INFO:teuthology.orchestra.run.smithi071.stdout:        "pid": "67436",
2024-02-28T08:14:43.775 INFO:teuthology.orchestra.run.smithi071.stdout:        "root": "/"
2024-02-28T08:14:43.775 INFO:teuthology.orchestra.run.smithi071.stdout:    },
2024-02-28T08:14:43.775 INFO:teuthology.orchestra.run.smithi071.stdout:    "dentry_count": 0,
2024-02-28T08:14:43.775 INFO:teuthology.orchestra.run.smithi071.stdout:    "dentry_pinned_count": 0,
2024-02-28T08:14:43.775 INFO:teuthology.orchestra.run.smithi071.stdout:    "id": 7134,
2024-02-28T08:14:43.776 INFO:teuthology.orchestra.run.smithi071.stdout:    "inst": {
2024-02-28T08:14:43.776 INFO:teuthology.orchestra.run.smithi071.stdout:        "name": {
2024-02-28T08:14:43.776 INFO:teuthology.orchestra.run.smithi071.stdout:            "type": "client",
2024-02-28T08:14:43.776 INFO:teuthology.orchestra.run.smithi071.stdout:            "num": 7134
2024-02-28T08:14:43.776 INFO:teuthology.orchestra.run.smithi071.stdout:        },
2024-02-28T08:14:43.776 INFO:teuthology.orchestra.run.smithi071.stdout:        "addr": {
2024-02-28T08:14:43.776 INFO:teuthology.orchestra.run.smithi071.stdout:            "type": "v1",
2024-02-28T08:14:43.776 INFO:teuthology.orchestra.run.smithi071.stdout:            "addr": "192.168.0.1:0",
2024-02-28T08:14:43.776 INFO:teuthology.orchestra.run.smithi071.stdout:            "nonce": 112361415
2024-02-28T08:14:43.776 INFO:teuthology.orchestra.run.smithi071.stdout:        }
2024-02-28T08:14:43.776 INFO:teuthology.orchestra.run.smithi071.stdout:    },
2024-02-28T08:14:43.776 INFO:teuthology.orchestra.run.smithi071.stdout:    "addr": {
2024-02-28T08:14:43.776 INFO:teuthology.orchestra.run.smithi071.stdout:        "type": "v1",
2024-02-28T08:14:43.777 INFO:teuthology.orchestra.run.smithi071.stdout:        "addr": "192.168.0.1:0",
2024-02-28T08:14:43.777 INFO:teuthology.orchestra.run.smithi071.stdout:        "nonce": 112361415
2024-02-28T08:14:43.777 INFO:teuthology.orchestra.run.smithi071.stdout:    },
2024-02-28T08:14:43.777 INFO:teuthology.orchestra.run.smithi071.stdout:    "inst_str": "client.7134 192.168.0.1:0/112361415",
2024-02-28T08:14:43.777 INFO:teuthology.orchestra.run.smithi071.stdout:    "addr_str": "192.168.0.1:0/112361415",
2024-02-28T08:14:43.777 INFO:teuthology.orchestra.run.smithi071.stdout:    "inode_count": 1,
2024-02-28T08:14:43.777 INFO:teuthology.orchestra.run.smithi071.stdout:    "mds_epoch": 32,
2024-02-28T08:14:43.777 INFO:teuthology.orchestra.run.smithi071.stdout:    "osd_epoch": 98,
2024-02-28T08:14:43.777 INFO:teuthology.orchestra.run.smithi071.stdout:    "osd_epoch_barrier": 0,
2024-02-28T08:14:43.777 INFO:teuthology.orchestra.run.smithi071.stdout:    "blocklisted": false,
2024-02-28T08:14:43.777 INFO:teuthology.orchestra.run.smithi071.stdout:    "fs_name": "cephfs"
2024-02-28T08:14:43.777 INFO:teuthology.orchestra.run.smithi071.stdout:}
```

but failed at writing a file:
```
2024-02-28T08:14:43.920 DEBUG:tasks.cephfs.mount:Creating test file 2024-02-28 08:14:43.920480 0
2024-02-28T08:14:43.920 DEBUG:teuthology.orchestra.run.smithi071:> touch '/home/ubuntu/cephtest/mnt.0/2024-02-28 08:14:43.920480 0'
2024-02-28T08:14:59.829 DEBUG:teuthology.orchestra.run.smithi071:> sudo logrotate /etc/logrotate.d/ceph-test.conf
2024-02-28T08:14:59.833 DEBUG:teuthology.orchestra.run.smithi164:> sudo logrotate /etc/logrotate.d/ceph-test.conf
<same logrotate logs till i killed the job>
```

**Comment by vshankar** [comment]
> looks like it did mount the client:
> [...]
> but failed at writing a file:
> 
> ```
> 2024-02-28T08:14:43.920 DEBUG:tasks.cephfs.mount:Creating test file 2024-02-28 08:14:43.920480 0
> 2024-02-28T08:14:43.920 DEBUG:teuthology.orchestra.run.smithi071:> touch '/home/ubuntu/cephtest/mnt.0/2024-02-28 08:14:43.920480 0'
> 2024-02-28T08:14:59.829 DEBUG:teuthology.orchestra.run.smithi071:> sudo logrotate /etc/logrotate.d/ceph-test.conf
> 2024-02-28T08:14:59.833 DEBUG:teuthology.orchestra.run.smithi164:> sudo logrotate /etc/logrotate.d/ceph-test.conf
> <same logrotate logs till i killed the job>
> ```

Have you checked if there is anything in the MDS logs for this request?

**Comment by vshankar** [comment]
Discussed with @gregsfortytwo today and since the blockage is during IO its very much possible that the new client doing IO is blocked on caps that's held by the other client which isn't evicted since an OSD is laggy.

**Comment by dparmar18** [comment]
> Discussed with @gregsfortytwo today and since the blockage is during IO its very much possible that the new client doing IO is blocked on caps that's held by the other client which isn't evicted since an OSD is laggy.

I was about to say the same, checked logs and yes this is the case:

mds log:
```
type=OPENED_INODES, session=0x55771942a500, opened_inodes=0, total_inodes=1
```

client log:
```
2024-03-01T09:01:21.842+0000 7f1b477f6640  1 -- 192.168.0.1:0/4001938067 --> [v2:172.21.15.120:6835/3304576025,v1:172.21.15.120:6837/3304576025] -- client_metrics [client_metric_type: READ_LATENCY latency: 0.000000, avg_latency: 0.000000, sq_sum: 0, count=0][client_metric_type: WRITE_LATENCY latency: 0.000000, avg_latency: 0.000000, sq_sum: 0, count=0][client_metric_type: METADATA_LATENCY latency: 0.000442, avg_latency: 0.000442, sq_sum: 0, count=1][client_metric_type: CAP_INFO cap_hits: 8 cap_misses: 2 num_caps: 0][client_metric_type: DENTRY_LEASE dlease_hits: 0 dlease_misses: 0 num_dentries: 0][client_metric_type: OPENED_FILES opened_files: 0 total_inodes: 1][client_metric_type: PINNED_ICAPS pinned_icaps: 1 total_inodes: 1][client_metric_type: OPENED_INODES opened_inodes: 0 total_inodes: 1][client_metric_type: READ_IO_SIZES total_ops: 0 total_size: 0][client_metric_type: WRITE_IO_SIZES total_ops: 0 total_size: 0] v1 -- 0x7f1b38004370 con 0x55f00b78a960
```

**Comment by dparmar18** [comment]
from mds log:

```
2024-03-01T09:07:16.545+0000 7fcac435f640  5 mds.0.server Detected 1 laggy clients, possibly due to laggy OSDs. Eviction is skipped until the OSDs return to normal.
2024-03-01T09:07:16.545+0000 7fcac435f640 10 mds.0.locker scatter_tick
2024-03-01T09:07:16.545+0000 7fcac435f640 20 mds.0.locker caps_tick 1 revoking caps
2024-03-01T09:07:16.545+0000 7fcac435f640 20 mds.0.locker caps_tick age = 3479.723644 client.6834.0x1
2024-03-01T09:07:16.545+0000 7fcac435f640 20 mds.0.locker caps_tick silencing log message (backoff) for client.6834.0x1
2024-03-01T09:07:16.545+0000 7fcac435f640 20 mds.0.bal handle_export_pins export_pin_queue size=0
2024-03-01T09:07:16.545+0000 7fcac435f640 15 mds.0.bal tick tick last_sample now 5763.148389s
2024-03-01T09:07:16.545+0000 7fcac435f640 15 mds.0.cache show_subtrees
2024-03-01T09:07:16.545+0000 7fcac435f640 10 mds.0.cache |__ 0    auth [dir 0x1 / [2,head] auth v=1 cv=1/1 REP dir_auth=0 ap=0+1 state=1073741825|complete f() n() hs=0+1,ss=0+0 | child=1 subtree=1 dirty=0 authpin=0 0x557719352480]
2024-03-01T09:07:16.545+0000 7fcac435f640 10 mds.0.cache |__ 0    auth [dir 0x100 ~mds0/ [2,head] auth v=1 cv=1/1 dir_auth=0 state=1073741825|complete f(v0 10=0+10) n(v0 rc2024-03-01T08:00:35.817341+0000 10=0+10) hs=10+0,ss=0+0 | child=1 subtree=1 dirty=0 authpin=0 0x557719352900]
2024-03-01T09:07:16.545+0000 7fcac435f640 10 mds.0.cache find_stale_fragment_freeze
2024-03-01T09:07:16.545+0000 7fcac435f640 10 mds.0.snap check_osd_map - version unchanged
2024-03-01T09:07:16.545+0000 7fcac435f640 20 mds.0.30 updating export targets, currently 0 ranks are targets
2024-03-01T09:07:16.545+0000 7fcac435f640 20 mds.beacon.a 1 slow request found
2024-03-01T09:07:16.833+0000 7fcac335d640 10 mds.pinger is_rank_lagging: rank=0
2024-03-01T09:07:16.833+0000 7fcac335d640  1 -- [v2:172.21.15.120:6835/3304576025,v1:172.21.15.120:6837/3304576025] --> [v2:172.21.15.120:6832/569419180,v1:172.21.15.120:6833/569419180] -- mgrreport(unknown.a +0-0 packed 1598) v9 -- 0x5577195c2e00 con 0x557719311000
2024-03-01T09:07:16.925+0000 7fcac8b68640  1 -- [v2:172.21.15.120:6835/3304576025,v1:172.21.15.120:6837/3304576025] <== client.7026 192.168.0.1:0/4001938067 3659 ==== client_metrics [client_metric_type: READ_LATENCY latency: 0.000000, avg_latency: 0.000000, sq_sum: 0, count=0][client_metric_type: WRITE_LATENCY latency: 0.000000, avg_latency: 0.000000, sq_sum: 0, count=0][client_metric_type: METADATA_LATENCY latency: 0.000442, avg_latency: 0.000442, sq_sum: 0, count=1][client_metric_type: CAP_INFO cap_hits: 8 cap_misses: 2 num_caps: 0][client_metric_type: DENTRY_LEASE dlease_hits: 0 dlease_misses: 0 num_dentries: 0][client_metric_type: OPENED_FILES opened_files: 0 total_inodes: 1][client_metric_type: PINNED_ICAPS pinned_icaps: 1 total_inodes: 1][client_metric_type: OPENED_INODES opened_inodes: 0 total_inodes: 1][client_metric_type: READ_IO_SIZES total_ops: 0 total_size: 0][client_metric_type: WRITE_IO_SIZES total_ops: 0 total_size: 0] v1 ==== 328+0+0 (crc 0 0 0) 0x5577192656c0 con 0x5577183db000
```

**Comment by dparmar18** [comment]
How about revoking the caps of the client that is laggy, renew the caps only after it becomes normal(when OSD(s) turn normal)? so if the new client joins and the old client becomes normal then we can renew the caps of the old client or maybe ask the client to reconnect?

**Comment by dparmar18** [comment]
> How about revoking the caps of the client that is laggy, provide back the caps only after it becomes normal(when OSD(s) turn normal)? if a new client joins and the old client becomes normal then we can renew the caps of the old client or maybe ask the client to reconnect?

@vshankar @gregsfortytwo do let me know your thoughts so that I start working on this ASAP

**Comment by vshankar** [comment]
> How about revoking the caps of the client that is laggy, provide back the caps only after it becomes normal(when OSD(s) turn normal)?

How is this different that what's being done right now? When the OSDs are not laggy anymore, the (existing) client would continue its operations as expected and the (new) client IO requests would be satisfied by revoking caps from the other client (if any).  

> if a new client joins and the old client becomes normal then we can renew the caps of the old client or maybe ask the client to reconnect?

Why reconnect?

**Comment by dparmar18** [comment]
> > How about revoking the caps of the client that is laggy, provide back the caps only after it becomes normal(when OSD(s) turn normal)?
> 

we're revoking client caps right now if a client is marked laggy due to laggy osds? i dont think so

> How is this different that what's being done right now? When the OSDs are not laggy anymore, the (existing) client would continue its operations as expected and the (new) client IO requests would be satisfied by revoking caps from the other client (if any).
> 
> > if a new client joins and the old client becomes normal then we can renew the caps of the old client or maybe ask the client to reconnect?
> 
> Why reconnect?

**Comment by vshankar** [comment]
> > > How about revoking the caps of the client that is laggy, provide back the caps only after it becomes normal(when OSD(s) turn normal)?
> 
> we're revoking client caps right now if a client is marked laggy due to laggy osds? i dont think so

That's not what I meant. The MDS revokes caps if required after then OSDs are not laggy and there is a contending client for the caps. If you mean to revoke caps from a client proactively when the OSDs are laggy and the MDS has chosen to not evict the client, the client could be unresponsive. What do you do in that case?

**Comment by dparmar18** [comment]
> If you mean to revoke caps from a client proactively when the OSDs are laggy and the MDS has chosen to not evict the client, the client could be unresponsive. What do you do in that case?

Circling back to the initial problem of client getting laggy due to laggy OSDs, what I understand is that there might be some I/O misses which asserts that I/O might be unsafe to do, so if a client is laggy(not nuked) then revoke the caps until the OSDs turn normal and pause the I/O, this means the client is just there, doing nothing. Any I/O op should return an errno to CLI with a err message saying "waiting for OSDs to get normal".

**Comment by vshankar** [comment]
> > If you mean to revoke caps from a client proactively when the OSDs are laggy and the MDS has chosen to not evict the client, the client could be unresponsive. What do you do in that case?
> 
> Circling back to the initial problem of client getting laggy due to laggy OSDs, what I understand is that there might be some I/O misses

I didn't get this part - IO misses? When as OSD is laggy, the client _could be_ blocked on flushing its buffered data out to the OSD.

> which asserts that I/O might be unsafe to do, so if a client is laggy(not nuked) then revoke the caps until the OSDs turn normal and pause the I/O,

What id an OSD becomes laggy and the client is unresponsive around the same time (due to network disruption or a client bug).

> this means the client is just there, doing nothing. Any I/O op should return an errno to CLI with an err message saying "waiting for OSDs to get normal".

Sorry, to which CLI?

**Comment by dparmar18** [comment]
> > > If you mean to revoke caps from a client proactively when the OSDs are laggy and the MDS has chosen to not evict the client, the client could be unresponsive. What do you do in that case?
> > 
> > 
> > Circling back to the initial problem of client getting laggy due to laggy OSDs, what I understand is that there might be some I/O misses
> 
> I didn't get this part - IO misses? 

if the osd is laggy, the comms b/w client and the data pool must have suffered i.e. the I/O should've also suffered

> When as OSD is laggy, the client _could be_ blocked on flushing its buffered data out to the OSD.

ok, but the "could" part should be _must_ is what im talking about, also what about other way around? what if the client wants some data from data pool i.e. retrieve/read? flushing data to the osd sounds we're trying to add something to data pool, but seems the fetch requests would be treated normally?

> 
> > which asserts that I/O might be unsafe to do, so if a client is laggy(not nuked) then revoke the caps until the OSDs turn normal and pause the I/O,
> 
> What if an OSD becomes laggy and the client is unresponsive around the same time (due to network disruption or a client bug).
> 

that is where the enhancement under https://tracker.ceph.com/issues/64563 will be done, currently all kind of unresponsiveness is considered as laggy, if a client is unable to renew caps in time it is considered laggy (which is wrong), this behaviour will be changed. so then you are mostly left with clients which are alive but laggy due to laggy osds, also the parameters to mark an osd laggy needs enhancement for which i've requested radasolow from the core team about this a week or two ago but no response yet.

> > this means the client is just there, doing nothing. Any I/O op should return an errno to CLI with an err message saying "waiting for OSDs to get normal".
> 
> Sorry, to which CLI?

to whatever interface the I/O is performed - report we're(client) waiting for osds to get normal and return EAGAIN or similar

**Comment by vshankar** [comment]
> > > > If you mean to revoke caps from a client proactively when the OSDs are laggy and the MDS has chosen to not evict the client, the client could be unresponsive. What do you do in that case?
> > > 
> > > 
> > > Circling back to the initial problem of client getting laggy due to laggy OSDs, what I understand is that there might be some I/O misses
> > 
> > 
> > I didn't get this part - IO misses?
> 
> if the osd is laggy, the comms b/w client and the data pool must have suffered i.e. the I/O should've also suffered

Right. "misses" is not the right word. It literally means that IOs are lost somewhat.

> 
> > When as OSD is laggy, the client _could be_ blocked on flushing its buffered data out to the OSD.
> 
> ok, but the "could" part should be _must_ is what im talking about, also what about other way around? what if the client wants some data from data pool i.e. retrieve/read? flushing data to the osd sounds we're trying to add something to data pool, but seems the fetch requests would be treated normally?
> 
> > > which asserts that I/O might be unsafe to do, so if a client is laggy(not nuked) then revoke the caps until the OSDs turn normal and pause the I/O,
> > 
> > 
> > What if an OSD becomes laggy and the client is unresponsive around the same time (due to network disruption or a client bug).
> 
> that is where the enhancement under https://tracker.ceph.com/issues/64563 will be done, currently all kind of unresponsiveness is considered as laggy, if a client is unable to renew caps in time it is considered laggy (which is wrong), this behaviour will be changed.

Hang on. Isn't this the same thing I proposed in https://github.com/ceph/ceph/pull/53578 - see, https://github.com/ceph/ceph/pull/53578#issuecomment-1761237445.

> so then you are mostly left with clients which are alive but laggy due to laggy osds, also the parameters to mark an osd laggy needs enhancement for which i've requested radasolow from the core team about this a week or two ago but no response yet.

So, basically we are talking about:

- enhancing OSD lagginess parameters
- inferring if a client is flushing out data to the osd which is laggy (rather than any osd being laggy)
- temporary (new) client blockage till the laggy OSD clears up or forcefully revoke caps from the laggy client

My argument now is that is this change worth doing for the problem it solves.

> 
> > > this means the client is just there, doing nothing. Any I/O op should return an errno to CLI with an err message saying "waiting for OSDs to get normal".
> > 
> > 
> > Sorry, to which CLI?
> 
> to whatever interface the I/O is performed - report we're(client) waiting for osds to get normal and return EAGAIN or similar

Returning -EAGAIN to clients is going to trip them off. The handling should be transparent to the client.

**Comment by dparmar18** [comment]
> > > When as OSD is laggy, the client _could be_ blocked on flushing its buffered data out to the OSD.
> > 
> > 
> > ok, but the "could" part should be _must_ is what im talking about, also what about other way around? what if the client wants some data from data pool i.e. retrieve/read? flushing data to the osd sounds we're trying to add something to data pool, but seems the fetch requests would be treated normally?
> > > > which asserts that I/O might be unsafe to do, so if a client is laggy(not nuked) then revoke the caps until the OSDs turn normal and pause the I/O,
> > > 
> > > 
> > > What if an OSD becomes laggy and the client is unresponsive around the same time (due to network disruption or a client bug).
> > 
> > 
> > that is where the enhancement under https://tracker.ceph.com/issues/64563 will be done, currently all kind of unresponsiveness is considered as laggy, if a client is unable to renew caps in time it is considered laggy (which is wrong), this behaviour will be changed.
> 
> Hang on. Isn't this the same thing I proposed in #53578 - see, [#53578 (comment)](https://github.com/ceph/ceph/pull/53578#issuecomment-1761237445).
> 

You're right there but I remember you wanted to remove the laggy clients detection completely from `Server::find_idle_session`, and that would be a bit too harsh, we don't know why client went silent, it could be anything right so I'd recommend we go only with:
https://github.com/ceph/ceph/blob/e913e5c6ce51753724e00491fab4de64502952f3/src/mds/Server.cc#L1184-L1190

but leave other check intact, this one:
https://github.com/ceph/ceph/blob/e913e5c6ce51753724e00491fab4de64502952f3/src/mds/Server.cc#L1200-L1217

because the second check depends on the client metadata, timeout there could be too less or just too much, also there is queue age involved, instead - just look at session autoclose which is part of MDS map which gives more control to MDS since the config is part of the MDS.

> > so then you are mostly left with clients which are alive but laggy due to laggy osds, also the parameters to mark an osd laggy needs enhancement for which i've requested radasolow from the core team about this a week or two ago but no response yet.
> 
> So, basically we are talking about:
> 
> * enhancing OSD lagginess parameters
> * inferring if a client is flushing out data to the osd which is laggy (rather than any osd being laggy)

if you feel this must be done then i'm good but I feel this would be a bit of an overkill. You mentioned in the above point to enhance OSD lagginess parameters then that should be respected and say hey there is an osd laggy so just wait for it be become normal, till then just revoke the client caps so that a) if the new client wants to joins then it is not blocked and b) if the old client comes back up(within the time interval of session autoclose since the session would be anyways evicted if it took more than this seconds to renew caps) then its caps can be renewed.

> * temporary (new) client blockage till the laggy OSD clears up or forcefully revoke caps from the laggy client

Check above point

> 
> My argument now is that is this change worth doing for the problem it solves.

yeah why not, this makes this whole detection code highly granular and generates less false positives.

> 
> > > > this means the client is just there, doing nothing. Any I/O op should return an errno to CLI with an err message saying "waiting for OSDs to get normal".
> > > 
> > > 
> > > Sorry, to which CLI?
> > 
> > 
> > to whatever interface the I/O is performed - report we're(client) waiting for osds to get normal and return EAGAIN or similar
> 
> Returning -EAGAIN to clients is going to trip them off. The handling should be transparent to the client.

Didn't quite get the transparent part

**Comment by dparmar18** [comment]
@vshankar ping

**Comment by vshankar** [comment]
> @vshankar ping

This got discussed a bit in yesterday's standup. @batrick has suggested an alternate approach involving a change in the protocol. I requested @batrick to share his thoughts in the tracker. Let's wait for his update (or else, I'll update as soon as I can).

**Comment by dparmar18** [comment]
> > @vshankar ping
> 
> I requested @batrick to share his thoughts in the tracker. Let's wait for his update (or else, I'll update as soon as I can).

this tracker https://tracker.ceph.com/issues/64563?

**Comment by vshankar** [comment]
> > > @vshankar ping
> > 
> > 
> > I requested @batrick to share his thoughts in the tracker. Let's wait for his update (or else, I'll update as soon as I can).
> 
> this tracker https://tracker.ceph.com/issues/64563?

yes

File: qa/tasks/cephfs/test_client_recovery.py
Comment (batrick) [inline qa/tasks/cephfs/test_client_recovery.py:901]: 6 minutes? `6*5` is ... 30 seconds?

File: qa/tasks/cephfs/test_client_recovery.py
Comment (batrick) [inline qa/tasks/cephfs/test_client_recovery.py:913]: With a suspended netns, do you expect this succeed? (I sure hope it doesn't...)

File: qa/tasks/cephfs/test_client_recovery.py
Comment (dparmar18) [reply qa/tasks/cephfs/test_client_recovery.py:901]: a few lines below:
```
                    with self.assert_cluster_log("1 client(s) laggy due to"
                                                 " laggy OSDs",
                                                 timeout=55):
```

total 6 iterations with sleep of 5 + 55

File: qa/tasks/cephfs/test_client_recovery.py
Comment (dparmar18) [reply qa/tasks/cephfs/test_client_recovery.py:913]: might be, i was hoping that the client should get blocklisted and this should take care of it:
https://github.com/ceph/ceph/blob/7d1f6d70f19b0144e31140a7887f5e2cd14d2dd0/qa/tasks/cephfs/fuse_mount.py#L321

plus the fact that it reaches `self.mount_a.create_destroy()` should mean the umount did went fine, right https://github.com/ceph/ceph/pull/55784#issuecomment-1968770427

**Comment by vshankar** [comment]
@dparmar18 gentle nudge on this - https://tracker.ceph.com/issues/64563#note-24

**Comment by vshankar** [comment]
> > @dparmar18 gentle nudge on this - https://tracker.ceph.com/issues/64563#note-24
> 
> as i mentioned on slack, design is fine, i need to start working on it but i need to get away with the client eviction code in find_idle_sessions that you once pushed in some of your PR. looks you're unavailable so i'll be doing the revert of the PRs laggy osds PRs #49971 #53839

Thanks. I'm on PTO next week too, so please take it up :)

File: qa/tasks/cephfs/test_client_recovery.py
Comment (batrick) [reply qa/tasks/cephfs/test_client_recovery.py:901]: the comment is still wrong?

File: qa/tasks/cephfs/test_client_recovery.py
Comment (batrick) [reply qa/tasks/cephfs/test_client_recovery.py:913]: That won't run because the client isn't blocklisted. I would expect this `umount_wait` to suspend until the MDS finally evicts the client after ~300 seconds.

In which case, I'm not sure what the point of this test is.

**Review by batrick — CHANGES_REQUESTED** [review]

**Pull request closed without merging**
